src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/ListBuilder.java

```diff
@@ -0,0 +1,121 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.javadoc.internal.doclets.formats.html.markup;
+
+import java.io.IOException;
+import java.io.Writer;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Objects;
+
+import jdk.javadoc.internal.doclets.formats.html.Content;
+
+/**
+ * A utility class for building nested HTML lists. This class is implemented as a
+ * stack of nested list elements where list items are added to the inner-most
+ * list. Lists can be added to and removed from the stack using the {@code pushNested}
+ * and {@code popNested} methods.
+ */
+public class ListBuilder extends Content {
+
+    private final HtmlTree root;
+    private final Deque<HtmlTree> stack = new ArrayDeque<>();
+    private HtmlTree currentList;
+    private HtmlTree currentItem;
+
+    /**
+     * Creates a new list builder.
+     *
+     * @param list the root list element
+     */
+    public ListBuilder(HtmlTree list) {
+        Objects.requireNonNull(list, "Root list element is required");
+        root = currentList = list;
+    }
+
+    /**
+     * Adds a new list item with the given content to the current list. The last added list item
+     * will also be used as container when pushing a new nested list.
+     *
+     * @param listItemContent the content for the new list item.
+     * @return this list builder
+     */
+    public ListBuilder add(Content listItemContent) {
+        currentItem = HtmlTree.LI(listItemContent);
+        currentList.addUnchecked(currentItem);
+        return this;
+    }
+
+    /**
+     * Adds a new nested list to the current list item without pushing it to the list stack and
+     * making it the recipient for new list items.
+     * @param list the nested list
+     * @return this list builder
+     */
+    public ListBuilder addNested(HtmlTree list) {
```

Comment:
General question: how aggressive do you want to make the checks? For example, do you want to make sure that lists are lists? (I hear @pavelrappo suggesting `assert` statements... (just kidding))